flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SlotSharingExecutionSlotAllocator.java

```diff
@@ -150,152 +167,72 @@ private SharedSlot getOrAllocateSharedSlot(
 		return sharedSlots
 			.computeIfAbsent(executionSlotSharingGroup, group -> {
 				SlotRequestId physicalSlotRequestId = new SlotRequestId();
+				ResourceProfile physicalSlotResourceProfile = getPhysicalSlotResourceProfile(group);
 				CompletableFuture<PhysicalSlot> physicalSlotFuture = sharedSlotProfileRetriever
-					.getSlotProfileFuture(group)
+					.getSlotProfileFuture(group, physicalSlotResourceProfile)
 					.thenCompose(slotProfile -> slotProvider.allocatePhysicalSlot(
 						new PhysicalSlotRequest(physicalSlotRequestId, slotProfile, slotWillBeOccupiedIndefinitely)))
 					.thenApply(PhysicalSlotRequest.Result::getPhysicalSlot);
-				return new SharedSlot(physicalSlotRequestId, group, physicalSlotFuture);
+				return new SharedSlot(
+					physicalSlotRequestId,
+					physicalSlotResourceProfile,
+					group,
+					physicalSlotFuture,
+					slotWillBeOccupiedIndefinitely,
+					this::releaseSharedSlot);
 			});
 	}
 
-	private class SharedSlot implements SlotOwner, PhysicalSlot.Payload {
-		private final SlotRequestId physicalSlotRequestId;
-
-		private final ExecutionSlotSharingGroup executionSlotSharingGroup;
-
-		private final CompletableFuture<PhysicalSlot> slotContextFuture;
-
-		private final DualKeyLinkedMap<ExecutionVertexID, SlotRequestId, CompletableFuture<SingleLogicalSlot>> requestedLogicalSlots;
-
-		private SharedSlot(
-				SlotRequestId physicalSlotRequestId,
-				ExecutionSlotSharingGroup executionSlotSharingGroup,
-				CompletableFuture<PhysicalSlot> slotContextFuture) {
-			this.physicalSlotRequestId = physicalSlotRequestId;
-			this.executionSlotSharingGroup = executionSlotSharingGroup;
-			this.slotContextFuture = slotContextFuture.thenApply(physicalSlot -> {
-				Preconditions.checkState(
-					physicalSlot.tryAssignPayload(this),
-					"Unexpected physical slot payload assignment failure!");
-				return physicalSlot;
-			});
-			this.requestedLogicalSlots = new DualKeyLinkedMap<>(executionSlotSharingGroup.getExecutionVertexIds().size());
-		}
-
-		private CompletableFuture<LogicalSlot> allocateLogicalSlot(ExecutionVertexID executionVertexId) {
-			Preconditions.checkArgument(executionSlotSharingGroup.getExecutionVertexIds().contains(executionVertexId));
-			CompletableFuture<SingleLogicalSlot> logicalSlotFuture = requestedLogicalSlots.getValueByKeyA(executionVertexId);
-			if (logicalSlotFuture != null) {
-				LOG.debug("Request for {} already exists", getLogicalSlotString(executionVertexId));
-			} else {
-				logicalSlotFuture = allocateNonExistentLogicalSlot(executionVertexId);
-			}
-			return logicalSlotFuture.thenApply(Function.identity());
+	private void releaseSharedSlot(ExecutionSlotSharingGroup executionSlotSharingGroup) {
+		SharedSlot slot = sharedSlots.remove(executionSlotSharingGroup);
```

Comment:
Maybe add an assertion that the `SharedSlot` is empty (has no requested logical slots assigned).